test/test_tensor_ops.py

```diff
@@ -0,0 +1,411 @@
+        elif device.type == 'cuda':
+            # bad CUDA Array Interface implementations
+            yield SampleInput(BadCudaArrayInterface1(), ref=ValueError('Missing __cuda_array_interface__'))
+            yield SampleInput(BadCudaArrayInterface2(), ref=ValueError('Invalid __cuda_array_interface__'))
+            yield SampleInput(BadCudaArrayInterface3(), ref=IndexError('bogus index error'))
+
+
+tensor_op_db = [TensorCreationOpInfo('as_tensor',
+                                     op=torch.as_tensor,
+                                     sample_inputs_func=sample_inputs_tensor,
+                                     supports_out=False),
+                TensorCreationOpInfo('tensor',
+                                     op=torch.tensor,
+                                     sample_inputs_func=sample_inputs_tensor,
+                                     supports_out=False),
+                TensorCreationOpInfo('tensor_pinning',
+                                     op=torch.tensor,
+                                     sample_inputs_func=functools.partial(sample_inputs_tensor, pin_memory=True),
+                                     supports_out=False),
+                TensorCreationOpInfo('tensor_grad',
+                                     op=torch.tensor,
+                                     sample_inputs_func=sample_inputs_tensor,
+                                     supports_autograd=True,
+                                     supports_out=False),
+                TensorCreationOpInfo('tensor_grad_pinning',
+                                     op=torch.tensor,
+                                     sample_inputs_func=functools.partial(sample_inputs_tensor, pin_memory=True),
+                                     supports_autograd=True,
+                                     supports_out=False)]
+
+
+class TestTensorCreationOps(TestCase):
+
+    def _check_shares_memory(self, tensor, other):
+        """Check if two tensors share memory.
+        """
+        # Notice that tensors may share memory even when using
+        # different devices: recall pin_memory option
+        if tensor.layout == torch.strided:
+            if tensor.storage().data_ptr() == other.storage().data_ptr():
+                return True
+            # TODO: check for strides
+        return False
+
+    def _apply_samples(self, device, dtype, op, test_func):
+        samples = op.sample_inputs(device, dtype, requires_grad=op.supports_autograd)
+        exceptions = []
+        count = 0
+        for sample in samples:
+            count += 1
+            ref = sample.ref
+            args = sample.args if sample.input is sample.unspecified else (sample.input,) + sample.args
+            assert sample.ref is not sample.unspecified
+            if callable(ref):
+                ref = ref(*args, **sample.kwargs)
+            try:
+                if isinstance(ref, Exception):
+                    assert len(ref.args) == 1
+                    try:
```

Comment:
I think the try/except is superfluous here?

Reply:
Not completely superfluous. The suggestion implementation gives a more informative error message. For instance, consider the case where `op` raises an exception that is different from `type(ref)`, then one gets the following error message:
```
  AssertionError:assert ValueError('b... index error') == IndexError('b... index error')
  +ValueError('bogus index error')
  -IndexError('bogus index error') [SampleInput(input=<BadCpuArrayInterface3 instance>, ref=bogus index error, extra={'sample-origin': '/home/pearu/git/pearu/pytorch/test/test_tensor_ops.py#297'})]
```
When eliminating the try/except block, one would get the following message:
```
IndexError:bogus index error [SampleInput(input=<BadCpuArrayInterface3 instance>, ref=bogus index error, extra={'sample-origin': '/home/pearu/git/pearu/pytorch/test/test_tensor_ops.py#297'})]
```
Notice that there is no information about the exception type (ValueError) that was actually raised.